### ISSUE TITLE
Default scheduling for matmul in nvFuser

### ISSUE THREAD

**Issue by Priya2698** [NVIDIA/Fuser#1669]
This issue is intended to track the development of default scheduling support for `matmul` and design options.

By default, whenever nvFuser encounters a matmul configuration that is not supported, we want it to be evaluated through the expression evaluator using PyTorch's `aten::matmul`.

CC: @protonu  @wujingyue @kevinstephano

**Comment by wujingyue** [comment]
@zasdfgbnm raised an important implementation challenge. There's apparently a plan to change nvFuser's interface so `fd.ops.matmul` will become `mul+sum` (and optionally broadcast and transpose) in the input FusionDefinition. The MatMulScheduler will be instead responsible to fuse that (or part of that) into an `MmaOp` with low-level details. This means `MmaOp` will only appear in MatMulScheduler::schedule and afterwards (e.g. device lowering). 

And thanks @zasdfgbnm for also explaining the motivations behind this plan of change:
1. MmaOp had been designed as a container for low level detail that user’s shouldn’t care. These details include Mma macros (like Ampere_16_8_16 ), layout (TT, TN, NT, NN), accumulator_stride  (a hack used to workaround a poor design in generated code, this hack no longer exist any more). It sounds weird to require users to provide these information. It makes more sense to have some simpler IR for the user to use, and let the matmul scheduler figure out these details and create the MmaOp.
2. The matmul scheduler is primarily for compute-bound matmuls. But depending on problem shape, matmul is not necessarily a compute bound kernel. In such senario, it would good to just let the matmul scheduler reject the fusion so that nvFuser can schedule it like a reduction, which can potentially fuse more ops.
3. At that time, there were no mechanism for nvFuser to invoke eager mode to evaluate tensor expressions. In that context, when users define a matmul, it is clearly a better option to schedule it like a reduction, compared with just fail and raise an error.
4. This design is kind of consistent with how we handle normalization: we do not have a NormalizationOp . Instead, normalization is defined as a reduction-broadcast-pointwise. Following the same scheme, it is natural to define matmul as mul-sum.

@protonu made a Monday meeting for us to talk about this. We'll clarify problems and hopefully converge to a solution that addresses the above concerns and this feature request.

**Comment by wujingyue** [comment]
Here are the meeting notes. cc attendees: @jacobhinkle , @kevinstephano , @Priya2698 , @protonu , @zasdfgbnm 

TL;DR: the recommendation so far is to pursue option 4 with option 5 as an intermediate step. 

# Background

To avoid a complicated dispatch in the framework, we want nvFuser to support compiling a matmul (however it’s represented) to one of the following and to automatically decide which path to take: 
1. an `at::matmul` call, a recent request from @kevinstephano to serve as the fallback option when the nvFuser-generated matmul kernel under performs.
2. a natively matmul-scheduled kernel. 
3. a natively reduction-scheduled kernel, which may be fastest in certain cases. 

Currently, nvFuser uses the following simple and suboptimal heuristics to decide which path to take:
1. nvFuser never takes path 1, because the integration doesn't yet exist. 
2. Whether path 2 or path 3 depends on how the matmul is represented. If it's represented as an MmaOp, it's picked up by MatmulScheduler; if ReductionOp, then ReductionScheduler. 

Very likely, we'll have to replace these with some scheduler-based heuristics, which have access to concrete shapes/strides. 

# Problems

1. Compile matmul to a `at::matmul` call. 
2. Support scheduler-based heuristics that decide the compilation path. 

Priority-wise, in O(month), we want nvFuser to support path 1 and 2 and to automatically decide between the two. After that, we may want to include path 3 to the equation. 

## Non-goals

Although not the focus of this discussion, we'll need to address other important problems, e.g., to optimize nvFuser schedulers to generate better kernels, and to come up with reasonable heuristics to decide which scheduler/kernel to use. 

# Proposed solutions

## Option 1

This is similar to how segmentation translates a WelfordOp today.

1. `fd.ops.matmul` builds BroadcastOp+MmaOp. 
2. MatmulScheduler rejects a MmaOp if it would be faster as a reduction kernel. 
3. Segmentation converts MmaOp to Mul+Sum if MatmulScheduler would reject the MmaOp. 

## Option 2

This is similar to option 1 but builds and converts BroadcastOp+Mul+Sum instead of MmaOp. 

1. `fd.ops.matmul` builds BroadcastOp+Mul+Sum. 
2. MatmulScheduler rejects a MmaOp if it would be faster as a reduction kernel. 
3. Segmentation converts Mul+Sum to MmaOp if MatmulScheduler would accept it. 

## ~Option 3~

1. `fd.ops.matmul` builds BroadcastOp+Mul+Sum. 
2. MatmulScheduler accepts a Mul+Sum pattern if it would be faster as a matmul kernel and converts the pattern to an MmaOp. 

This doesn't work because ReductionScheduler will instead take over BroadcastOp+Mul+Sum. To fix #1628, we want `MatmulScheduler` to reject the BroadcastOp to avoid an unnecessary shared-to-shared memcpy around tmaLoad. Even if we make MatmulScheduler accept Mul+Sum, the larger pattern Broadcast+Mul+Sum will be accepted by ReductionScheduler, because segmentation prefers larger segments. 

## Option 4

Ignore path 3 for now. I.e., decision between path 2 and path 3 remains to be op-type (MmaOp or ReductionOp) based, which is status quo. 

1. `fd.ops.matmul` builds BroadcastOp+MmaOp
2. If MatmulScheduler thinks an MmaOp is faster as an `at::matmul`, it will skip normal scheduling and simply mark the fusion to compute the MmaOp using `ExpressionEvaluator`. 
3. Teach device lowering to skip such MmaOps. 
4. Teach `FusionExecutor` to compute such MmaOps using `ExpressionEvaluator`. 

In terms of implementation, we'll try to reuse the existing alias infra and make it more general. The above flow is similar to how `markAliases` marks pointer-arithmetic aliases via `Fusion::aliasOutputToInput`. Device lowering already [skips Exprs](https://github.com/NVIDIA/Fuser/blob/cad93e5163b587d661c443b51f04408d3f5b996e/csrc/device_lower/pass/expr_sort.cpp#L1518
) that are only depended by aliases, and `allocateOutputs` already calls `ExpressionEvaluator` to compute aliases. 

## Option 5 (proposed by @protonu)

As an intermediate step of option 4, manually (via some API) mark the fusion to compute the MmaOp via `ExpressionEvaluator`. This allows us to integrate `MmaOp::evaluate` in parallel to finding a good heuristic. 

# Recommendation so far

Implementation option 4 with option 5 as an intermediate step. It's simple to implement and unblocks the near-term need of supporting path 1 and path 2.

**Comment by jacobhinkle** [comment]
> How do we plan to compile a `half,half->float` `MmaOp` to a `half,half->half` `at::matmul`?

That would probably be tough, but because of this PyTorch behavior of always preserving the input dtype in `mm`, we can probably translate to `MmaOp + castOp` in integration, then we should never encounter such a "HTH" matmul in the wild.

**Comment by wujingyue** [comment]
> we can probably translate to MmaOp + castOp in integration

Yes, we can. Still, we wouldn't be able to expression-evaluate an `MmaOp` by its own. I'd be surprised if PyTorch doesn't have an API to do a half-to-float matmul. Isn't that commonly used by these attention models to preserve high precision before softmax? An I missing something?

**Comment by zasdfgbnm** [comment]
What would we get from python frontend? Will we ever get a HH->S matmul from our frontend, considering that PyTorch does not support it? If we always get a HH->H matmul from frontend, then likely we will also always get fusions like `MmaOp->cast`. If we naively uses ExpressionEvaluator to evaluate this fusion, it will not be fast, because it will be translated to:
```C++
at::matmul(a, b).to(kFloat).to(kHalf)
```

IIUC, our plan for invoking `at::matmul` is to let `scheduleMatmul` skip the scheduling part and set some special flag to the fusion to request eagerly evaluate? Could we do the following?

```C++
void scheduleMatmul(Fusion* fusion, const MatmulParams& params) {
  if (params.uses_eager_mode) {
    // convert cast(H, MmaOp(in1=H, in2=H, out=S)) into MmaOp(in1=H, in2=H, out=H)
    fusion.eagerly_evaluate = true;
    return;
  }
  // schedule matmul as usual.
}
```

**Comment by wujingyue** [comment]
> What would we get from python frontend?

Good question -- it seems complicated... 

Focusing on MLP for now, `torch.nn.Linear` becomes a `linear` primitive in the framework. We haven't registered any rule yet to translate that primitive to nvFuser, so it's up for us to decide... 

If we want the `linear` primitive to follow the same behavior as PyTorch, the matmul there should be half-to-half [^1]. However, there seems to be [a backend flag](https://pytorch.org/docs/stable/notes/numerical_accuracy.html#reduced-precision-reduction-for-fp16-and-bf16-gemms) controlling the **compute** data type. So I'm not sure an `at::matmul` is equivalent to MmaOp+Cast.

[^1] In PyTorch, `torch.nn.linear` becomes `torch.nn.functional.linear`, which then becomes [this C++ implementation](https://github.com/pytorch/pytorch/blob/697ca4f2921a9939c54eb3f9167386783f13c7c7/torch/csrc/api/include/torch/nn/functional/linear.h#L19). It calls `at::Tensor::matmul` (when bias is not defined) or `torch::addmm` (when bias is defined), which AFAICT does half-to-half. 

If we can customize (can we? @kevinstephano), we'll likely translate it to something convenient for our implementation. 

> If we naively uses ExpressionEvaluator to evaluate this fusion, it will not be fast

That's right. 

Question about MatmulScheduler: do we plan to fuse output pointwise ops (e.g. biasadd and relu) into the matmul kernel? I don't see our current implementation does that, but I'd guess so. That would make it even harder to efficiently expression-evaluate the matmul segment, because the segment would look like `MmaOp=>Add=>Relu=>Cast`.

**Comment by jacobhinkle** [comment]
> There seems to be [a backend flag](https://pytorch.org/docs/stable/notes/numerical_accuracy.html#reduced-precision-reduction-for-fp16-and-bf16-gemms) controlling the **compute** data type. So I'm not sure an `at::matmul` is equivalent to MmaOp+Cast. 

That flag controls the type of the reduction buffer for splitk, enabling reduced precision for that partial reduction. The output type of the matmul should be unaffected by the flag though, as I understand it. I will test it this morning.

**Comment by zasdfgbnm** [comment]
> I don't see our current implementation does that, but I'd guess so.

We actually do support epilogue fusions:
https://github.com/NVIDIA/Fuser/blob/main/test/test_matmul_scheduler.cpp

**Comment by Priya2698** [comment]
We support ATen evaluation for matmul and linear now, closing this issue.
